pingcap/tidb
# *: fix cancel query slow_query doesn't release resource and tiny optimize

### What problem does this PR solve?

Same problem with log-search in https://github.com/pingcap/tidb/pull/21373

![image](https://user-images.githubusercontent.com/26020263/100961265-0100ea80-355d-11eb-8d83-b6d011d080c7.png)


1. Check `ctx.Done` as early as possible. When the query slow_log is canceled, the slow log parse-worker should exist as early as possible to release the resource.
2. optimize the readLastLines function for huge lines.

Here is a benchmark when the last log line size is `10MB`:

```go
Before this PR:
searchLogSuite.BenchmarkReadLastLinesOfHugeLine          1        10227299446 ns/op

This PR:
searchLogSuite.BenchmarkReadLastLinesOfHugeLine        100          14824774 ns/op
```

### What is changed and how it works?

N/A

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test
   1. run an expensive query of `cluster_slow_query`.
   2. cancel the query.
   3. check the goroutine whether contain any goroutine that belong to `slow_query.go`

Side effects

- No

### Release note 

- Fix issue of canceled query slow_query doesn't release resource and tiny optimize
